x/concentrated-liquidity/incentives.go

```diff
@@ -193,6 +193,94 @@ func (k Keeper) prepareBalancerPoolAsFullRange(ctx sdk.Context, clPoolId uint64)
+	// Get CL pool from ID. This also serves as an early pool existence check.
+	clPool, err := k.getPoolById(ctx, clPoolId)
+	if err != nil {
+		return sdk.Coins{}, err
+	}
+
+	// Get longest lockup period for pool
+	lockableDurations := k.poolIncentivesKeeper.GetLockableDurations(ctx)
+	longestLockableDuration := lockableDurations[len(lockableDurations)-1]
+
+	// Get gauge corresponding to the longest lockup period
+	gaugeId, err := k.poolIncentivesKeeper.GetPoolGaugeId(ctx, balPoolId, longestLockableDuration)
+	if err != nil {
+		return sdk.Coins{}, err
+	}
+
+	// Get all uptime accumulators for CL pool
+	// Create a temporary position record on all uptime accumulators with this amount. We expect this to be cleared later
+	// with `claimAndResetFullRangeBalancerPool`
+	uptimeAccums, err := k.getUptimeAccumulators(ctx, clPoolId)
+	if err != nil {
+		return sdk.Coins{}, err
+	}
+
+	// Claim rewards on each uptime accumulator. Delete each record after claiming.
+	totalRewards := sdk.NewCoins()
+	for uptimeIndex, uptimeAccum := range uptimeAccums {
+		// Generate key for the record on the the current uptime accumulator
+		balancerPositionName := string(types.KeyBalancerFullRange(clPoolId, balPoolId, uint64(uptimeIndex)))
+
+		// Ensure that the given balancer pool has a record on the given uptime accumulator.
+		// We expect this to have been set in a prior call to `prepareBalancerAsFullRange`, which
+		// should precede all calls of `claimAndResetFullRangeBalancerPool`
+		recordExists, err := uptimeAccum.HasPosition(balancerPositionName)
+		if err != nil {
+			return sdk.Coins{}, err
+		}
+		if !recordExists {
+			return sdk.Coins{}, types.BalancerRecordNotFoundError{ClPoolId: clPoolId, BalancerPoolId: balPoolId, UptimeIndex: uint64(uptimeIndex)}
+		}
+
+		// Remove shares from record so it gets cleared when rewards are claimed
+		numShares, err := uptimeAccum.GetPositionSize(balancerPositionName)
+		if err != nil {
+			return sdk.Coins{}, err
+		}
+
+		err = uptimeAccum.RemoveFromPosition(balancerPositionName, numShares)
+		if err != nil {
+			return sdk.Coins{}, err
+		}
+
+		// Claim rewards and log the amount claimed to be added to the relevant gauge later
+		claimedRewards, _, err := uptimeAccum.ClaimRewards(balancerPositionName)
+		if err != nil {
+			return sdk.Coins{}, err
+		}
+		totalRewards = totalRewards.Add(claimedRewards...)
+
+		// Ensure record was deleted
+		recordExists, err = uptimeAccum.HasPosition(balancerPositionName)
+		if err != nil {
+			return sdk.Coins{}, err
+		}
+		if recordExists {
+			return sdk.Coins{}, types.BalancerRecordNotClearedError{ClPoolId: clPoolId, BalancerPoolId: balPoolId, UptimeIndex: uint64(uptimeIndex)}
+		}
+	}
+
+	// After claiming accrued rewards from all uptime accumulators, add the total claimed amount to the
+	// Balancer pool's longest duration gauge.
+	err = k.incentivesKeeper.AddToGaugeRewards(ctx, clPool.GetIncentivesAddress(), totalRewards, gaugeId)
```

Comment:
Let's surround this with `if !totalRewards.IsZero()` please.

I'm seeing that there are hooks that run "after adding to gauge". Would be good to avoid triggering listeners redundantly when nothing actually gets added to gauge

Reply:
Good point!